### PR TITLE
x11-wm/fvwm3: new package, add 1.0.4, 9999

### PR DESCRIPTION
**Pull request by Kangie**
Needle in IRC asked if someone could give bug 760012 some attention, so I took a look and updated the work done by @domichel 

Speaking of, happy to throw you down as the proxied maintainer @domichel - I use KDE most of the time!

Patches for -9999 have been reapplied and cleaned, new patches backported for 1.0.4. I haven't had a chance to thoroughly test yet but it's good enough for a review.

I chose to go with upstream's naming convention - fvwm3, rather than adding a new version to `x11-wm/fvwm` as it simplifies updates and management - upstream numbering is fvwm3-1.0.4 currently. At the very least I'll need to update the PR to add `x11-wm/fvwm` as a blocker (I'll do that now...), but I imagine that anything that currently depends on `x11-wm/fvwm` should get an additional dependency on this.

I'll take some time later this week to test more thoroughly; right now it compiles and installs but I have no idea if the transparency patches actually do anything; they descended from some patches for fvwm in ~2003 on the mailing list by the looks of things!

Any suggestions / feedback please let me know and I'll get it addressed ASAP

Closes: https://bugs.gentoo.org/760012

Signed-off-by: Matt Jolly <Matt.Jolly@footclan.ninja>

### CODE REVIEW

**Comment by Kangie** [comment]
Also, I'm using  `EGO_SUM` here as I don't have a devspace to upload a tarball. If anyone is able to help out with that I'm all ears.

**Comment by Kangie** [comment]
OK, that should have fixed the QA Errors for `EGO_SUM`.

**Comment by Kangie** [comment]
Can confirm that 1.0.4 seems to work on X11; suspect that someone that actually uses fvwm3 will need to go through ebuilds that depend on fvwm and add ` || x11-wm\fvwm3`, but I don't want to go ahead and apply that across the board as there's likely to be exceptions.

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:40]: Libraries which are linked against are usually for DEPEND + RDEPEND, not BDEPEND.

BDEPEND is for tools which get executed (on CBUILD) during, well, the build, so stuff like unzip, Go.

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:57]: := on deps with subslots that we link against (both ncurses and readline)

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:83]: Everything from pkgconfig down (inclusive) except for xorg-proto == BDEPEND

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:96]: || die on external commands which may fail

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:99]: Put stuff like this above all phases, i.e. move src_unpack down below all of this

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:116]: It'd be better to do this before eautoreconf, but may need to sed Makefile.am or something instead if they're using automake.

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:126]: "live code"?

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:130]: This makes me think: we should maybe have a flag-o-matic thing for this, or tc-has-unsigned-char function. Anyway, you need to add in arm and arm64 here. I think ppc64 too.

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:136]: convert to array please

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:150]: missing die

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:148]: use edo instead? but if it's autotools, use econf

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:156]: use htmldoc && emake -C doc html

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:165]: emake, not make, then you can drop the die

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-1.0.4.ebuild:168]: ${ED}

File: x11-wm/fvwm3/fvwm3-9999.ebuild
Comment (thesamesam) [inline x11-wm/fvwm3/fvwm3-9999.ebuild:6]: 8..11

File: x11-wm/fvwm3/metadata.xml
Comment (thesamesam) [inline x11-wm/fvwm3/metadata.xml:23]: We don't have Gentoo/FreeBSD anymore. Drop.

File: x11-wm/fvwm3/metadata.xml
Comment (thesamesam) [inline x11-wm/fvwm3/metadata.xml:18]:
```suggestion
		<flag name="go">Enable building <pkg>dev-lang/go</pkg> code</flag>
```
but ideally explain what it actually does

File: x11-wm/fvwm3/metadata.xml
Comment (thesamesam) [inline x11-wm/fvwm3/metadata.xml:17]: maybe just USE=doc for this?

**Review by thesamesam — CHANGES_REQUESTED** [review]

**Comment by thesamesam** [comment]
and nice, thanks for working on this!

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (Kangie) [reply x11-wm/fvwm3/fvwm3-1.0.4.ebuild:126]: From the context, I imagine that means 'live ebuilds', but I also can't see it hurting anything...

Comment was already in the attached bug ebuild that I adapted.

File: x11-wm/fvwm3/fvwm3-1.0.4.ebuild
Comment (Kangie) [reply x11-wm/fvwm3/fvwm3-1.0.4.ebuild:130]: That would be a useful helper.

**Comment by Kangie** [comment]
QA thinks that RDEPEND is wrong, any thoughts?

```
 * VDB: detected possibly incorrect RDEPEND (x11-wm/fvwm3-1.0.4)
 *                     > dev-libs/glib:2
 *                     > dev-libs/libevent:=
 * sys-libs/ncurses:=  < 
 * sys-libs/readline:= < 
 * sys-libs/zlib       < 
 *                     > x11-libs/cairo
 * x11-libs/libXau     < 
 * x11-libs/libXdmcp   < 
 * x11-libs/libXfixes  < 
 * x11-libs/libxcb     < 
 ```

Otherwise I think that I've addressed all of your feedback, let me know if there's anything else that can be done to improve the commit.

**Comment by thesamesam** [comment]
It seems to link against readline for some stuff but maybe the resultant binaries don't get installed?

```
[...]
x86_64-pc-linux-gnu-gcc  -Wall -Wno-implicit-int -fno-common -Wno-error=format-security -O2 -pipe -march=native -fdiagnostics-color=always -frecord-gcc-switches     -ggdb3 -Werror=format-security -Werror=implicit-function-declaration -Wformat -fno-strict-aliasing -I/usr/include/freetype2 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include   -Wl,-O1 -Wl,--as-needed -Wl,--defsym=__gentoo_check_ldflags__=0 -o FvwmConsoleC FvwmConsoleC.o getline.o -L../../libs -lfvwm3 -lreadline -lncurses -lfreetype   -lXft  -lfontconfig -lfreetype
x86_64-pc-linux-gnu-gcc  -Wall -Wno-implicit-int -fno-common -Wno-error=format-security -O2 -pipe -march=native -fdiagnostics-color=always -frecord-gcc-switches     -ggdb3 -Werror=format-security -Werror=implicit-function-declaration -Wformat -fno-strict-aliasing -I/usr/include/freetype2 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include   -Wl,-O1 -Wl,--as-needed -Wl,--defsym=__gentoo_check_ldflags__=0 -o FvwmConsole FvwmConsole.o -L../../libs -lfvwm3 -lreadline -lncurses -lfreetype   -lXft  -lfontconfig -lfreetype
[...]
```

Note that iwdevtools for the RDEPEND check can only show linked-against deps, not "data deps" (like a dictionary), or binaries which get called at runtime, etc, or dlopen(). But I don't think that's the case here anyway.

I guess I'd keep readline+ncurses in for now. All that other stuff seems linked against too.

**Comment by thesamesam** [comment]
Didn't mean to push, sorry!

**Comment by Kangie** [comment]
All good.

Re: FvwmConsole, looks like it's not always installed, according to the note in postinst - seems to depend on use=go but I don't see any logic for that. Seems harmless enough and something that I'm happy to revisit after a version bump.

Fixed some deps, should be good to go unless you can spot any showstoppers. Outside of that I can't think of anything unless there's a preference to lump this in with `x11-wm/fvwm`. A quick look at repology suggests that fvwm3 is considered a separate package in other distros so I'm happy to leave it as-is.

Quick sanity check for packages that depend on fvwm v2 and might need to be updated after this is merged:

```
x11-themes/fvwm-crystal/fvwm-crystal-3.6.5.ebuild
x11-themes/fvwm-crystal/fvwm-crystal-3.7.1.ebuild
x11-themes/fvwm-themes/fvwm-themes-0.7.0-r1.ebuild
x11-themes/fvwm_icons/fvwm_icons-1.0.ebuild
x11-themes/fvwm_sounds/fvwm_sounds-1.0.ebuild
```

Most of the changes will probably just be `|| ( x11-wm/fvwm3 x11-wm/fvwm )` in depend and a quick validation that it actually works.

I'm probably not going to have a chance to look at any of the remaining packages (though the bumps seem trivial). I have added a new commit to bump `x11-themes/fvwm-crystal` and add the fvwm3 dep as it's supposed to be compatible.

@domichel Your website, sourceforge repo, and git repo all point at different versions of fvwm-crystal, FYI.

**Comment by Kangie** [comment]
I updated (and EAPI bumped) fvwm_icons and fvwm_sounds. I do note that `||` prefers the leftmost option, but this shouldn't impact users until fvwm3 is stablised, and even then will only impact new users (and frankly, if you somehow manage to install a window manager from an icons package you're probably using portage in a particularly interesting way). Happy to change that if we'd prefer fvwm be on the left.

`fvwm-themes` also hasn't had a code change in 19 years; I'm not touching that - if an `fvwm3` user decides that they need the functionality that can be revisited, but it may (or may not) work with fvwm3.

**Pull request closed without merging**
